src/reply_manager.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! A mixnode may receive the same request multiple times due to retransmission (see eg
+//! [`request_manager`](super::request_manager)). A [`ReplyManager`] can be used to cache replies,
+//! to avoid needing to execute requests more than once.
+
+use super::core::{MessageId, Mixnet, RequestMessage, SessionIndex, Surb, MESSAGE_ID_SIZE};
+use hashlink::{linked_hash_map::Entry, LinkedHashMap};
+use log::{debug, warn};
+use rand::RngCore;
+use std::time::{Duration, Instant};
+
+/// Reply manager configuration.
+#[derive(Clone, Debug)]
+pub struct Config {
+	/// The target for log messages.
+	pub log_target: &'static str,
+	/// Maximum number of requests to remember. When this limit is reached, old requests will be
+	/// automatically discarded to make space for new ones.
+	pub capacity: usize,
+	/// Maximum number of copies of a reply message to post in response to a single request
+	/// message. Note that the number of copies is also limited by the number of SURBs provided in
+	/// the request message.
+	pub max_posts: usize,
```

Comment:
It is the number of simultaneous replies yes. I tried to make that clear, not really sure how to word it. As noted though, it's only really a limit because if there aren't enough SURBs we might not be able to post all the replies. I don't know if it is theoretically possible to have replicable SURBs, but the current implementation does not support anything like this; you just need to make sure you include enough SURBs in requests.

Not sure merging the reply and request managers really makes sense; the reply manager can't really retry even if it has spare SURBs because it doesn't know if the replies it sent were received or not. The retrying is really just driven by the requester, with this multiple-posts thing wedged on a bit at the reply side because it's so easy to do.